src/opera_utils/_utils.py

```diff
@@ -153,3 +161,238 @@ def scratch_directory(
+        lat_lon_radar = np.array(
+            transformer_xy_to_latlon.TransformPoints(x_y_pnts_radar)
+        )
+
+        snwe = (
+            lat_lon_radar[0, 0],
+            lat_lon_radar[1, 0],
+            lat_lon_radar[0, 1],
+            lat_lon_radar[1, 1],
+        )
+    else:
+        snwe = (bounds[1], bounds[3], bounds[0], bounds[2])
+
+    return snwe
+
+
+def create_yx_arrays(
+    gt: list[float], shape: tuple[int, int]
+) -> tuple[np.ndarray, np.ndarray]:
+    """Create the x and y coordinate datasets.
+
+    Parameters
+    ----------
+    gt : List[float]
+        Geotransform list.
+    shape : Tuple[int, int]
+        Shape of the dataset (ysize, xsize).
+
+    Returns
+    -------
+    Tuple[np.ndarray, np.ndarray]
+        x and y coordinate arrays.
+    """
+    ysize, xsize = shape
+    # Parse the geotransform
+    x_origin, x_res, _, y_origin, _, y_res = gt
+    y_end = y_origin + y_res * ysize
+    x_end = x_origin + x_res * xsize
+
+    # Make the x/y arrays
+    y = np.arange(y_origin, y_end - 500, -500)
+    x = np.arange(x_origin, x_end + 500, 500)
+    return y, x
+
+
+def transform_xy_to_latlon(
+    epsg: int, x: ArrayLike, y: ArrayLike
+) -> Tuple[ArrayLike, ArrayLike]:
+    """Convert the x, y coordinates in the source projection to WGS84 lat/lon.
+
+    Parameters
+    ----------
+    epsg : int
+        EPSG code.
+    x : ArrayLike
+        x coordinates.
+    y : ArrayLike
+        y coordinates.
+
+    Returns
+    -------
+    Tuple[ArrayLike, ArrayLike]
+        Latitude and longitude arrays.
+    """
+    # x, y to Lat/Lon
+    srs_src = osr.SpatialReference()
+    srs_src.ImportFromEPSG(epsg)
+
+    srs_wgs84 = osr.SpatialReference()
+    srs_wgs84.ImportFromEPSG(4326)
+
+    if epsg != 4326:
+        # Transform the xy to lat/lon
+        transformer_xy_to_latlon = osr.CoordinateTransformation(srs_src, srs_wgs84)
+
+        # Stack the x and y
+        x_y_pnts_radar = np.stack((x.flatten(), y.flatten()), axis=-1)
+
+        # Transform to lat/lon
+        lat_lon_radar = np.array(
+            transformer_xy_to_latlon.TransformPoints(x_y_pnts_radar)
+        )
+
+        # Lat lon of data cube
+        lat_datacube = lat_lon_radar[:, 0].reshape(x.shape)
+        lon_datacube = lat_lon_radar[:, 1].reshape(x.shape)
+    else:
+        lat_datacube = y.copy()
+        lon_datacube = x.copy()
+
+    ## Extent of the data cube
+    # cube_extent = (np.nanmin(lat_datacube) - margin, np.nanmax(lat_datacube) + margin,
+    #               np.nanmin(lon_datacube) - margin, np.nanmax(lon_datacube) + margin)
+
+    return lat_datacube, lon_datacube  # , cube_extent
+
+
+def compute_2d_delay(
+    tropo_delay_cube: dict, grid: dict, geo_files: dict[str, Path]
+) -> dict:
+    """Compute 2D delay.
+
+    Parameters
+    ----------
+    tropo_delay_cube : dict
+        Dictionary containing tropospheric delay data.
+    grid : dict
+        Dictionary containing grid information.
```

Comment:
for the `grid` and `tropo_delay_cube` input2, a couple thoughts:

1. Maybe we can make then into `dataclass` instead of `dict`s. That way, a user could tell what is needed to run `compute_2d_delay` just from the input types.

(Assuming I understood correctly that it's getting made in the dolphin PR)
So instead of 
```python
    grid = {'xcoord': xcoord,
            'ycoord': ycoord,
            'height_levels': tropo_height_levels,
            'snwe': oput.get_snwe(epsg, out_bounds),
            'epsg': epsg,
            'geotransform': gt,
            'shape': (ysize, xsize),
            'crs': crs.to_wkt()} # cube
```

```python
from dataclasses import dataclass

@dataclass
class DelayParams:
    x_coordinates: np.ndarray
    y_coordinates: np.ndarray
    bounds: Bbox
    crs: pyproj.CRS  # (or str, or maybe making this `epsg: int`)
    ...
```

2. Since it's nice to have functions which run with as few required inputs as possible, lets figure out what we definitely need here- It looks like we're not using the `snwe` or the `epsg` (since we're using `crs`), so let's cut those. Also we could pass the `bounds` instead of recomputing from `gt`, but that's up to you

3. Do you think we'll want to use this anywhere except in the dolphin function?  I only ask because right now it's a little hard to tell how the inputs could get formed without some context that comes from the dolphin function... If we think it'll be nice to have in a couple repos besides dolphin, we could keep this here and add the `dataclass` which describes the input here

Reply:
Thank you @scottstanie 
I agree with all the changes. I have moved the compute_2d_delay function to dolphin for now since it is only used for troposphere and changed dict to dataclass. Please take a look when possible